File: js/types/src/submission.ts
import { Literal, Partial, Record, Static, String, Union } from "runtypes"
import { Languages } from "./languages"
import { TestTestingSettings } from "./testtesting"

export const SubmissionType = Union(Literal("SOLVE"), Literal("TESTTESTING"))
export type SubmissionType = Static<typeof SubmissionType>

export const Submission = Record({
  type: SubmissionType,
  contentHash: String,
  language: Languages,
  contents: String,
}).And(
  Partial({
    testTestingSettings: TestTestingSettings,
  }),
)
export type Submission = Static<typeof Submission>

File: lib/src/main/kotlin/server/Submission.kt
package edu.illinois.cs.cs125.questioner.lib.server

import edu.illinois.cs.cs125.questioner.lib.Language
import edu.illinois.cs.cs125.questioner.lib.Question
import kotlinx.serialization.Serializable

@Serializable
data class Submission(
    val type: SubmissionType,
    val contentHash: String,
    val language: Language,
    val contents: String,
    val originalID: String? = null,
    val testTestingSettings: Question.TestTestingSettings? = null,
) {
    enum class SubmissionType { SOLVE, TESTTESTING }
}

fun Question.toSubmission(type: Submission.SubmissionType, language: Language, contents: String) =
    Submission(type, published.contentHash, language, contents)